projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,280 @@
+
+The wallet storage will be initialized with the `xPub` ( reference: [`initHWStorage`](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/storage.js#L127-L140) function ) obtained from the HSM: a read-only wallet. On the 2nd phase, signatures are made with the [`ledger/sendTx()` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/utils/ledger.js#L126-L166), as described on the Headless section above.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Multiple HSM devices
+The proposed structure enforces that only one HSM device can be accessed by a headless application environment, since the authentication credentials are stored on config variables.
+
+Another approach to this would be to transfer these credentials to the `/hsm/start` requests, allowing each `walletId` to store its own credentials besides the HSM key itself.
+
+The downside of this solution is that the security of these credentials would be in responsibility of the end user, instead of wrapped within the security of the Headless Wallet configurations.
+
+# Prior art
+[prior-art]: #prior-art
+
+A similar solution to this integration was implemented on the Desktop Wallet for interacting with the Ledger Hardware Wallet. This design uses many of the concepts and code implementations to achieve a similar objective.
+
+The main difference between the Ledger and the HSM is that here there is no need for human intervention on the steps.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The only question left is if our application is able to correctly sign a transaction with the HSM in its current state, or if changes will have to be made to the Hathor Lib or the Dinamo-Hsm lib to allow this integration. The 2nd phase will be dependent on this answer, and it is being discussed on [internal issue 205](https://github.com/HathorNetwork/internal-issues/issues/205).
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Full integration on all headless endpoints
+As explained in the [headless implementation section](#headless-implementation), only the atomic swap `get-my-signatures` endpoint was adapted to interact with the HSM.
+
+Adapting all endpoints to use the HSM would require more profound refactorings, more specifically the [utils/transaction#prepareTransaction()](https://github.com/HathorNetwork/hathor-wallet-lib/blob/d3dbe159ac121eb67986ed8561cdacead8fc9fe8/src/utils/transaction.ts#L485-L502) method. From the perspective of the headless wallet, all signatures are executed through this method.
+
+Below are the endpoints that would need to be integrated with the HSM ( [link to the router file](https://github.com/HathorNetwork/hathor-wallet-headless/blob/1fc34b6f12a349478d9656b14933e6db45115a90/src/routes/wallet/wallet.routes.js) for conference ):
```

Comment:
You don't need to add specific support for each of these endpoints. We can add support to build any type of tx, then the signature/mining/push of the tx is the same for all of them.

Reply:
The final conclusion of this section of the document was also that: we could add support to a specific method on the wallet lib and compatibility with HSM signatures would become automatic.

On [ce01489](https://github.com/HathorNetwork/rfcs/pull/67/commits/ce01489eec9c6976213f825171471b06764b8c04) removed the section that contained the research part and left only the final conclusion of where the adaptation will be needed on the wallet lib.